File: debrid/torbox.js
const axios = require("axios");
const https = require("https");

const TB_BASE = "https://api.torbox.app/v1/api";
const TB_TIMEOUT = 60000;
const LIST_CACHE_TTL = 30000;
const MAX_RETRIES = 4;
const MAX_INFO_POLLS = 4;
const POLL_DELAYS = [1200, 1800, 2500, 3500];

const VIDEO_EXTENSIONS = /\.(mkv|mp4|avi|mov|wmv|flv|webm|iso|m4v|ts)$/i;
const JUNK_PATTERN = /\b(sample|trailer|promo|preview|screens?|proof|nfo|cover|poster|thumb|trailerfix)\b/i;
const EXTRA_PATTERN = /\b(ova|oad|special|extras?|featurette|intervista|interview|behind\s*the\s*scenes|making\s*of|recap|riassunto|ncop|nced|creditless|ost|soundtrack)\b/i;
const RESETTABLE_STATES = new Set(["error", "metadl", "stalled"]);
const COMPLETED_STATES = new Set(["completed", "seeding", "ready"]);

const PUBLIC_TRACKERS = [
  "udp://tracker.opentrackr.org:1337/announce",
  "udp://open.demonii.com:1337/announce",
  "udp://tracker.coppersurfer.tk:6969/announce",
  "udp://tracker.leechers-paradise.org:6969/announce",
  "udp://9.rarbg.to:2710/announce",
  "udp://tracker.openbittorrent.com:80/announce",
  "udp://opentracker.i2p.rocks:6969/announce"
];

const httpsAgent = new https.Agent({
  keepAlive: true,
  rejectUnauthorized: process.env.TB_INSECURE_SSL === "1" ? false : true,
  maxSockets: 50
});

const COMMON_HEADERS = {
  "User-Agent": "Leviathan/2.0 (TorBoxModule)",
  Accept: "application/json"
};

const userListCache = new Map();

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function lower(value) {
  return String(value || "").trim().toLowerCase();
}

function safeInt(value, fallback = 0) {
  const parsed = Number.parseInt(value, 10);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function safeNum(value, fallback = 0) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function stableKeyFingerprint(key) {
  const str = String(key || "");
  if (!str) return "__empty__";
  if (str.length <= 10) return str;
  return `${str.slice(0, 4)}:${str.slice(-4)}`;
}

function getCacheEntry(key) {
  if (!userListCache.has(key)) {
    userListCache.set(key, { data: null, timestamp: 0, pending: null });
  }
  return userListCache.get(key);
}

function invalidateUserList(key) {
  if (key) {
    userListCache.delete(key);
    return;
  }
  userListCache.clear();
}

function pruneExpiredCache() {
  const now = Date.now();
  for (const [key, entry] of userListCache.entries()) {
    if (!entry.pending && (!entry.timestamp || now - entry.timestamp > LIST_CACHE_TTL * 3)) {
      userListCache.delete(key);
    }
  }
}

function getRetryDelay(error, attempt) {
  const status = error?.response?.status;
  const retryAfter = safeInt(error?.response?.headers?.["retry-after"]);
  if (retryAfter > 0) return retryAfter * 1000;
  if (status === 429) return 4000 + attempt * 1200;
  return 1200 * (attempt + 1);
}

function isRetryableError(error) {
  const status = error?.response?.status;
  if (status === 408 || status === 425 || status === 429) return true;
  if (status >= 500) return true;
  return ["ECONNABORTED", "ECONNRESET", "ETIMEDOUT", "EAI_AGAIN", "ENOTFOUND"].includes(error?.code);
}

async function tbRequest(method, endpoint, key, { data = null, params = null, timeout = TB_TIMEOUT } = {}) {
  let lastResponse = null;
  for (let attempt = 0; attempt < MAX_RETRIES; attempt += 1) {
    try {
      const config = {
        method,
        url: `${TB_BASE}${endpoint}`,
        headers: { ...COMMON_HEADERS, Authorization: `Bearer ${key}` },
        timeout,
        params,
        httpsAgent,
        validateStatus: () => true
      };

      if (method === "POST" && data) {
        const form = new URLSearchParams();
        for (const [k, v] of Object.entries(data)) {
          if (v == null) continue;
          form.append(k, String(v));
        }
        config.data = form;
        config.headers["Content-Type"] = "application/x-www-form-urlencoded";
      }

      const response = await axios(config);
      lastResponse = response;
      if (response.status >= 200 && response.status < 300) {
        return response;
      }
      if (attempt < MAX_RETRIES - 1 && isRetryableError({ response })) {
        await sleep(getRetryDelay({ response }, attempt));
        continue;
      }
      return response;
    } catch (error) {
      if (attempt < MAX_RETRIES - 1 && isRetryableError(error)) {
        await sleep(getRetryDelay(error, attempt));
        continue;
      }
      return error?.response || lastResponse || null;
    }
  }
  return lastResponse;
}

async function fetchUserList(key, extraParams = null) {
  const res = await tbRequest("GET", "/torrents/mylist", key, {
    params: { bypass_cache: true, ...(extraParams || {}) }
  });
  if (!res?.data?.data) return null;
  const data = Array.isArray(res.data.data) ? res.data.data.slice() : [res.data.data];
  data.sort((a, b) => safeInt(b?.id) - safeInt(a?.id));
  return data;
}

async function getUserList(key, forceRefresh = false) {
  pruneExpiredCache();
  const entry = getCacheEntry(key);
  const age = Date.now() - entry.timestamp;

  if (!forceRefresh && entry.data && age < LIST_CACHE_TTL) {
    return entry.data;
  }
  if (entry.pending) {
    return entry.pending;
  }

  entry.pending = (async () => {
    const list = await fetchUserList(key);
    if (list) {
      entry.data = list;
      entry.timestamp = Date.now();
    }
    entry.pending = null;
    return list;
  })();

  return entry.pending;
}

function normalizeName(value) {
  return lower(value)
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[\[\](){}]/g, " ")
    .replace(/[_+.]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function getFileName(file) {
  return String(file?.name || file?.short_name || "");
}

function getFileId(file) {
  const raw = file?.id ?? file?.file_id ?? null;
  const parsed = safeInt(raw, NaN);
  return Number.isFinite(parsed) ? parsed : null;
}

function getFileSize(file) {
  return Math.max(0, safeNum(file?.size, 0));
}

function isVideoFile(file) {
  const name = getFileName(file);
  return VIDEO_EXTENSIONS.test(name) && !JUNK_PATTERN.test(name);
}

function isExtraFile(file) {
  return EXTRA_PATTERN.test(getFileName(file));
}

function isSeasonPackName(name, season) {
  if (!season) return false;
  return new RegExp(`\bseason\s*0*${season}\b|\bcomplete\b|\bpack\b|\b全集\b|\bcollection\b|\bintegrale\b|\bstagione\s*0*${season}\b`, "i").test(name);
}

function parseEpisodeRange(name) {
  const patterns = [
    /\bE?(\d{1,3})\s*[-~]\s*E?(\d{1,3})\b/i,
    /\b(?:episodes?|eps?)\s*(\d{1,3})\s*[-~]\s*(\d{1,3})\b/i
  ];
  for (const pattern of patterns) {
    const match = name.match(pattern);
    if (!match) continue;
    const start = safeInt(match[1], 0);
    const end = safeInt(match[2], 0);
    if (start > 0 && end >= start) {
      return { start, end };
    }
  }
  return null;
}

function buildEpisodeRegexes(season, episode) {
  const s = safeInt(season, 0);
  const e = safeInt(episode, 0);
  const e2 = String(e).padStart(2, "0");
  return [
    { score: 1200, regex: new RegExp(`\bS(?:eason)?\s*0*${s}\s*[-_. ]*E(?:pisode)?\s*0*${e}\b`, "i") },
    { score: 1180, regex: new RegExp(`\b0*${s}x0*${e}\b`, "i") },
    { score: 1140, regex: new RegExp(`\bS0*${s}[^a-z0-9]{0,4}E0*${e}\b`, "i") },
    { score: 1090, regex: new RegExp(`\b${s}${e2}\b`) },
    { score: 980, regex: new RegExp(`\bepisode\s*0*${e}\b|\bep\s*0*${e}\b`, "i") },
    { score: 760, regex: new RegExp(`(?:^|\D)0*${e}(?:\D|$)`) }
  ];
}

function buildMovieScore(file) {
  const name = normalizeName(getFileName(file));
  const size = getFileSize(file);
  let score = 0;
  if (isVideoFile(file)) score += 400;
  score += Math.min(size / (1024 * 1024 * 40), 180);
  if (isExtraFile(file)) score -= 300;
  if (/\b(sample|trailer|extras?|featurette|making of)\b/i.test(name)) score -= 350;
  if (/\b(2160p|1080p|720p|bluray|bdrip|web[- ]?dl|webrip|remux)\b/i.test(name)) score += 25;
  return score;
}

function buildEpisodeScore(file, season, episode) {
  const name = normalizeName(getFileName(file));
  const size = getFileSize(file);
  let score = 0;

  if (!isVideoFile(file)) return -Infinity;

  score += 300;
  score += Math.min(size / (1024 * 1024 * 50), 150);
  if (isExtraFile(file)) score -= 320;

  for (const { score: bonus, regex } of buildEpisodeRegexes(season, episode)) {
    if (regex.test(name)) {
      score += bonus;
      break;
    }
  }

  const range = parseEpisodeRange(name);
  const targetEpisode = safeInt(episode, 0);
  if (range) {
    if (targetEpisode >= range.start && targetEpisode <= range.end) {
      score -= 220;
    } else {
      score -= 500;
    }
  }

  if (isSeasonPackName(name, season)) score -= 180;
  if (/\b(batch|multi|全集|complete|collection|pack)\b/i.test(name)) score -= 120;
  if (/\b(2160p|1080p|720p|bluray|bdrip|web[- ]?dl|webrip|remux)\b/i.test(name)) score += 20;

  return score;
}

function resolveForcedFileId(files, forcedFileIdx) {
  if (forcedFileIdx == null) return null;
  const raw = safeInt(forcedFileIdx, NaN);
  if (!Number.isFinite(raw) || raw < 0) return null;

  const byId = files.find((file) => getFileId(file) === raw);
  if (byId) return getFileId(byId);

  if (raw < files.length) {
    return getFileId(files[raw]);
  }

  return null;
}

function chooseBestFile(files, season, episode) {
  const list = Array.isArray(files) ? files.slice() : [];
  if (list.length === 0) return null;

  const videoFiles = list.filter((file) => isVideoFile(file));
  const candidates = videoFiles.length > 0 ? videoFiles : list;

  const s = safeInt(season, 0);
  const e = safeInt(episode, 0);
  const isMovie = s <= 0 && e <= 0;

  const ranked = candidates
    .map((file) => ({
      file,
      score: isMovie ? buildMovieScore(file) : buildEpisodeScore(file, s, e),
      size: getFileSize(file)
    }))
    .filter((item) => Number.isFinite(item.score))
    .sort((a, b) => (b.score - a.score) || (b.size - a.size));

  return ranked[0]?.file || null;
}

function matchFile(files, season, episode, forcedFileIdx = null) {
  const list = Array.isArray(files) ? files : [];
  if (list.length === 0) return null;

  const forced = resolveForcedFileId(list, forcedFileIdx);
  if (forced != null) return forced;

  const selected = chooseBestFile(list, season, episode);
  return selected ? getFileId(selected) : null;
}

function extractHash(magnet) {
  const value = String(magnet || "");
  const match = value.match(/btih:([a-zA-Z0-9]{32,40})/i);
  return match ? match[1].toLowerCase() : null;
}

function enhanceMagnet(magnet) {
  const raw = String(magnet || "").trim();
  if (!raw) return raw;
  const lowerMagnet = raw.toLowerCase();
  const missing = PUBLIC_TRACKERS.filter((tracker) => !lowerMagnet.includes(encodeURIComponent(tracker).toLowerCase()) && !lowerMagnet.includes(tracker.toLowerCase()));
  if (missing.length === 0) return raw;
  return `${raw}${missing.map((tracker) => `&tr=${encodeURIComponent(tracker)}`).join("")}`;
}

async function deleteTorrent(key, torrentId) {
  const res = await tbRequest("POST", "/torrents/controltorrent", key, {
    data: { torrent_id: torrentId, operation: "delete" }
  });
  invalidateUserList(key);
  return Boolean(res?.data?.success || (res?.status >= 200 && res?.status < 300));
}

function shouldResetTorrent(torrent) {
  const state = lower(torrent?.download_state);
  const progress = safeNum(torrent?.progress, 0);
  return RESETTABLE_STATES.has(state) && progress < 100;
}

function pickDeletionVictim(list) {
  if (!Array.isArray(list) || list.length === 0) return null;

  const sortOldest = (items) => items
    .slice()
    .sort((a, b) => new Date(a?.created_at || 0).getTime() - new Date(b?.created_at || 0).getTime())[0] || null;

  const completed = list.filter((t) => COMPLETED_STATES.has(lower(t?.download_state)));
  if (completed.length > 0) return sortOldest(completed);

  const resettable = list.filter((t) => RESETTABLE_STATES.has(lower(t?.download_state)));
  if (resettable.length > 0) return sortOldest(resettable);

  return null;
}

async function freeUpSpace(key) {
  const list = await getUserList(key, true);
  const victim = pickDeletionVictim(list);
  if (!victim?.id) return false;
  const ok = await deleteTorrent(key, victim.id);
  if (ok) await sleep(900);
  return ok;
}

async function findTorrentByHash(key, hash) {
  if (!hash) return null;
  const list = await getUserList(key);
  return list?.find((torrent) => lower(torrent?.hash) === lower(hash)) || null;
}

async function refreshTorrentInfo(key, torrentId) {
  if (!torrentId) return null;
  const list = await fetchUserList(key, { id: torrentId });
  if (!list || list.length === 0) return null;
  return list.find((torrent) => safeInt(torrent?.id) === safeInt(torrentId)) || list[0] || null;
}

async function waitForFiles(key, torrentId, initialFiles = null) {
  if (Array.isArray(initialFiles) && initialFiles.length > 0) return initialFiles;
  for (let i = 0; i < MAX_INFO_POLLS; i += 1) {
    await sleep(POLL_DELAYS[i] || POLL_DELAYS[POLL_DELAYS.length - 1]);
    const info = await refreshTorrentInfo(key, torrentId);
    if (Array.isArray(info?.files) && info.files.length > 0) return info.files;
    if (COMPLETED_STATES.has(lower(info?.download_state)) && Array.isArray(info?.files)) {
      return info.files;
    }
  }
  return initialFiles;
}

async function createTorrent(key, magnet) {
  const postData = { magnet, seed: "1", allow_zip: "false" };
  let createRes = await tbRequest("POST", "/torrents/createtorrent", key, { data: postData });

  const detail = createRes?.data?.detail || createRes?.data?.error || "";
  const detailStr = typeof detail === "object" ? JSON.stringify(detail) : String(detail || "");

  if (!createRes?.data?.success && /limit|active|storage|space/i.test(detailStr)) {
    const freed = await freeUpSpace(key);
    if (freed) {
      createRes = await tbRequest("POST", "/torrents/createtorrent", key, { data: postData });
    }
  }

  return createRes;
}

async function requestDownloadLink(key, torrentId, fileId, userIp = null) {
  const params = {
    token: key,
    torrent_id: torrentId,
    file_id: fileId,
    zip_link: "false"
  };
  if (userIp) params.user_ip = userIp;
  return tbRequest("GET", "/torrents/requestdl", key, { params });
}

const TB = {
  checkCached: async (key, hashes) => {
    const clean = Array.from(new Set((Array.isArray(hashes) ? hashes : [])
      .map((hash) => lower(hash))
      .filter(Boolean)));
    if (clean.length === 0) return [];

    const res = await tbRequest("GET", "/torrents/checkcached", key, {
      params: {
        hash: clean.join(","),
        format: "object",
        list_files: "false"
      },
      timeout: 20000
    });

    if (!res?.data?.data) return [];

    const cached = [];
    for (const [hash, info] of Object.entries(res.data.data)) {
      if (info?.cached || (Array.isArray(info?.files) && info.files.length > 0) || safeNum(info?.size, 0) > 0) {
        cached.push(lower(hash));
      }
    }
    return cached;
  },

  getStreamLink: async (key, magnet, season = null, episode = null, hash = null, forcedFileIdx = null, userIp = null) => {
    try {
      let torrentId = null;
      let files = null;
      const targetHash = lower(hash || extractHash(magnet));
      const enhancedMagnet = enhanceMagnet(magnet);

      if (targetHash) {
        const existing = await findTorrentByHash(key, targetHash);
        if (existing) {
          if (shouldResetTorrent(existing)) {
            console.log(`🧹 [TorBox:${stableKeyFingerprint(key)}] reset torrent ${existing.id} state=${lower(existing.download_state)}`);
            await deleteTorrent(key, existing.id);
            await sleep(800);
          } else {
            torrentId = existing.id;
            files = existing.files;
          }
        }
      }

      if (!torrentId) {
        const createRes = await createTorrent(key, enhancedMagnet);

        if (!createRes?.data?.success) {
          const detail = createRes?.data?.detail || createRes?.data?.error || "Unknown";
          const detailStr = typeof detail === "object" ? JSON.stringify(detail) : String(detail);

          if (/exists/i.test(detailStr) && targetHash) {
            const existing = await findTorrentByHash(key, targetHash);
            if (existing) {
              torrentId = existing.id;
              files = existing.files;
            }
          }

          if (!torrentId) {
            throw new Error(`Add Failed: ${detailStr}`);
          }
        } else {
          torrentId = createRes.data.data?.torrent_id || createRes.data.data?.id || null;
          files = createRes.data.data?.files || null;
          invalidateUserList(key);
        }
      }

      if (!torrentId) {
        throw new Error("Torrent ID not available");
      }

      files = await waitForFiles(key, torrentId, files);
      const fileId = matchFile(files, season, episode, forcedFileIdx);

      if (fileId == null) {
        throw new Error("File not found inside torrent");
      }

      let linkRes = await requestDownloadLink(key, torrentId, fileId, userIp);

      if ((!linkRes?.data?.success || !linkRes?.data?.data) && (!files || files.length === 0)) {
        files = await waitForFiles(key, torrentId, files);
        const retryFileId = matchFile(files, season, episode, forcedFileIdx);
        if (retryFileId != null && retryFileId !== fileId) {
          linkRes = await requestDownloadLink(key, torrentId, retryFileId, userIp);
        }
      }

      if (linkRes?.data?.success && linkRes.data.data) {
        return { url: linkRes.data.data };
      }

      const detail = linkRes?.data?.detail || linkRes?.data?.error || `HTTP ${linkRes?.status || "unknown"}`;
      const detailStr = typeof detail === "object" ? JSON.stringify(detail) : String(detail);
      throw new Error(`Link Request Failed: ${detailStr}`);
    } catch (error) {
      console.error(`💥 [TorBox:${stableKeyFingerprint(key)}] ${error?.message || error}`);
      return null;
    }
  }
};

module.exports = TB;

File: core/torbox.py
from __future__ import annotations

import asyncio
import logging
import os
from typing import Any

import httpx

from core.app_settings import TORBOX_POLL_ATTEMPTS, TORBOX_POLL_INTERVAL
from core.config import coerce_int
from core.rd import build_magnet_link, rank_file_candidates, select_best_file
from core.stream_pipeline import TRACKERS

logger = logging.getLogger("torrenthan.torbox")
TORBOX_API_BASE = "https://api.torbox.app/v1/api"
TORBOX_REQUESTDL_RETRIES = coerce_int(os.getenv("TORBOX_REQUESTDL_RETRIES"), 3, minimum=1, maximum=10)
TORBOX_FILE_CANDIDATES = coerce_int(os.getenv("TORBOX_FILE_CANDIDATES"), 4, minimum=1, maximum=12)
TORBOX_MYLIST_LIMIT = coerce_int(os.getenv("TORBOX_MYLIST_LIMIT"), 1000, minimum=1, maximum=5000)


def _clean_api_key(api_key: str) -> str:
    return str(api_key or "").replace("Bearer ", "").strip()


def torbox_headers(api_key: str) -> dict[str, str]:
    clean_key = _clean_api_key(api_key)
    return {
        "Authorization": f"Bearer {clean_key}",
        "Accept": "application/json",
        "User-Agent": "torrenthan-torbox/5.1",
    }


def _response_json(response: httpx.Response) -> dict[str, Any]:
    try:
        data = response.json()
    except Exception:
        return {"success": False, "error": response.text[:500], "data": None}
    return data if isinstance(data, dict) else {"success": False, "error": "invalid_json_shape", "data": data}


async def torbox_create_torrent(client: httpx.AsyncClient, api_key: str, magnet: str) -> dict[str, Any]:
    response = await client.post(
        f"{TORBOX_API_BASE}/torrents/createtorrent",
        data={"magnet": magnet, "seed": "1", "allow_zip": "false"},
        headers=torbox_headers(api_key),
    )
    if response.status_code >= 400:
        logger.warning("TorBox createtorrent %s - %s", response.status_code, response.text[:350])
        return {"success": False, "error": response.text[:500], "data": None}
    return _response_json(response)


async def torbox_list_torrents(client: httpx.AsyncClient, api_key: str) -> list[dict[str, Any]]:
    response = await client.get(
        f"{TORBOX_API_BASE}/torrents/mylist",
        params={"bypass_cache": "true", "limit": TORBOX_MYLIST_LIMIT},
        headers=torbox_headers(api_key),
    )
    if response.status_code >= 400:
        logger.warning("TorBox mylist %s - %s", response.status_code, response.text[:350])
        return []
    data = _response_json(response).get("data", [])
    return data if isinstance(data, list) else []


async def torbox_get_torrent(client: httpx.AsyncClient, api_key: str, torrent_id: str | int) -> dict[str, Any] | None:
    if str(torrent_id or "").strip() == "":
        return None
    response = await client.get(
        f"{TORBOX_API_BASE}/torrents/mylist",
        params={"bypass_cache": "true", "id": torrent_id},
        headers=torbox_headers(api_key),
    )
    if response.status_code >= 400:
        logger.warning("TorBox mylist?id=%s %s - %s", torrent_id, response.status_code, response.text[:350])
        return None
    data = _response_json(response).get("data")
    if isinstance(data, dict):
        return data
    if isinstance(data, list) and data:
        found = next((item for item in data if str(item.get("id")) == str(torrent_id)), None)
        return found or data[0]
    return None


def _extract_torrent_id(create_response: dict[str, Any]) -> str | None:
    data = create_response.get("data") if isinstance(create_response, dict) else None
    if isinstance(data, dict):
        for key in ("torrent_id", "id"):
            value = str(data.get(key) or "").strip()
            if value:
                return value
    if isinstance(data, int):
        return str(data)
    if isinstance(data, str) and data.strip().isdigit():
        return data.strip()
    return None


def _torrent_matches_hash(item: dict[str, Any], hash_val: str) -> bool:
    needle = str(hash_val or "").strip().lower()
    if not needle:
        return False
    if str(item.get("hash") or "").strip().lower() == needle:
        return True
    alternatives = item.get("alternative_hashes") or []
    if isinstance(alternatives, list):
        return any(str(value or "").strip().lower() == needle for value in alternatives)
    return False


def _torrent_files(item: dict[str, Any] | None) -> list[dict[str, Any]]:
    if not isinstance(item, dict):
        return []
    files = item.get("files") or []
    return [file_item for file_item in files if isinstance(file_item, dict)] if isinstance(files, list) else []


def _normalize_torbox_files(files: list[dict[str, Any]]) -> list[dict[str, Any]]:
    mapped: list[dict[str, Any]] = []
    for idx, file_item in enumerate(files):
        file_id = file_item.get("id")
        if file_id is None or str(file_id).strip() == "":
            continue
        path = (
            file_item.get("absolute_path")
            or file_item.get("short_name")
            or file_item.get("name")
            or file_item.get("path")
            or ""
        )
        mapped.append(
            {
                "id": str(file_id).strip(),
                "path": str(path or ""),
                "name": str(file_item.get("name") or path or ""),
                "short_name": str(file_item.get("short_name") or ""),
                "bytes": file_item.get("size", 0),
                "size": file_item.get("size", 0),
                "mimetype": str(file_item.get("mimetype") or ""),
                "_original_index": idx,
            }
        )
    return mapped


def _candidate_file_ids(
    mapped_files: list[dict[str, Any]],
    *,
    season: int,
    episode: int,
    file_index: str,
) -> list[str]:
    candidates: list[str] = []

    selected = select_best_file(
        mapped_files,
        season=season,
        episode=episode,
        file_index_hint=file_index,
    )
    if selected:
        candidates.append(str(selected))

    ranked = rank_file_candidates(
        mapped_files,
        season=season,
        episode=episode,
        file_index_hint=file_index,
        limit=TORBOX_FILE_CANDIDATES,
    )
    for file_id in ranked:
        value = str(file_id or "").strip()
        if value and value not in candidates:
            candidates.append(value)

    return candidates[:TORBOX_FILE_CANDIDATES]


async def torbox_request_download(
    client: httpx.AsyncClient,
    api_key: str,
    torrent_id: str | int,
    file_id: str,
) -> str | None:
    clean_key = _clean_api_key(api_key)
    for attempt in range(TORBOX_REQUESTDL_RETRIES):
        response = await client.get(
            f"{TORBOX_API_BASE}/torrents/requestdl",
            params={
                "token": clean_key,
                "torrent_id": torrent_id,
                "file_id": file_id,
                "zip_link": "false",
                "append_name": "true",
            },
            headers=torbox_headers(api_key),
        )
        if response.status_code >= 400:
            logger.warning(
                "TorBox requestdl failed | status=%s | torrent=%s | file=%s | attempt=%s/%s | body=%s",
                response.status_code,
                torrent_id,
                file_id,
                attempt + 1,
                TORBOX_REQUESTDL_RETRIES,
                response.text[:300],
            )
            if response.status_code in {408, 425, 429, 500, 502, 503, 504} and attempt < TORBOX_REQUESTDL_RETRIES - 1:
                await asyncio.sleep(min(3.0, 0.8 * (attempt + 1)))
                continue
            return None

        payload = _response_json(response)
        if not payload.get("success"):
            logger.warning(
                "TorBox requestdl unsuccessful | torrent=%s | file=%s | detail=%s | error=%s",
                torrent_id,
                file_id,
                str(payload.get("detail") or "")[:180],
                str(payload.get("error") or "")[:180],
            )
            return None

        data = payload.get("data")
        if isinstance(data, str) and data.strip():
            return data.strip()
        if isinstance(data, dict):
            for key in ("download", "link", "url"):
                value = str(data.get(key) or "").strip()
                if value:
                    return value
        logger.warning("TorBox requestdl returned no URL | torrent=%s | file=%s", torrent_id, file_id)
        return None
    return None


async def _find_torbox_torrent(
    client: httpx.AsyncClient,
    api_key: str,
    hash_val: str,
    torrent_id_hint: str | None,
) -> dict[str, Any] | None:
    if torrent_id_hint:
        target = await torbox_get_torrent(client, api_key, torrent_id_hint)
        if target and (not hash_val or _torrent_matches_hash(target, hash_val) or str(target.get("id")) == str(torrent_id_hint)):
            return target

    torrents = await torbox_list_torrents(client, api_key)
    return next((item for item in torrents if _torrent_matches_hash(item, hash_val)), None)


async def resolve_torbox_download(
    hash_val: str,
    api_key: str,
    season: int,
    episode: int,
    file_index: str,
    client: httpx.AsyncClient,
) -> str | None:
    try:
        full_magnet = build_magnet_link(hash_val, trackers=TRACKERS)
    except ValueError:
        logger.warning("TorBox invalid hash for playback | hash=%s", str(hash_val)[:80])
        return None

    try:
        create_response = await torbox_create_torrent(client, api_key, full_magnet)
        torrent_id_hint = _extract_torrent_id(create_response)
        if not create_response.get("success") and not torrent_id_hint:
            logger.warning(
                "TorBox create failed | hash=%s | detail=%s | error=%s",
                hash_val,
                str(create_response.get("detail") or "")[:180],
                str(create_response.get("error") or "")[:180],
            )

        target: dict[str, Any] | None = None
        raw_files: list[dict[str, Any]] = []
        for attempt in range(TORBOX_POLL_ATTEMPTS):
            target = await _find_torbox_torrent(client, api_key, hash_val, torrent_id_hint)
            raw_files = _torrent_files(target)
            if target and raw_files:
                break
            if attempt < TORBOX_POLL_ATTEMPTS - 1:
                await asyncio.sleep(TORBOX_POLL_INTERVAL)

        if not target:
            logger.warning("TorBox torrent not found after create | hash=%s | torrent_id=%s", hash_val, torrent_id_hint or "n/a")
            return None

        torrent_id = target.get("id") or torrent_id_hint
        if not torrent_id:
            logger.warning("TorBox target without torrent id | hash=%s", hash_val)
            return None

        if not raw_files:
            logger.warning(
                "TorBox torrent has no files yet | hash=%s | torrent=%s | state=%s | progress=%s | cached=%s",
                hash_val,
                torrent_id,
                target.get("download_state"),
                target.get("progress"),
                target.get("cached"),
            )
            return None

        mapped_files = _normalize_torbox_files(raw_files)
        candidate_ids = _candidate_file_ids(mapped_files, season=season, episode=episode, file_index=file_index)
        if not candidate_ids:
            logger.warning(
                "TorBox no playable candidate | hash=%s | torrent=%s | files=%s",
                hash_val,
                torrent_id,
                len(mapped_files),
            )
            return None

        for selected_file_id in candidate_ids:
            final_url = await torbox_request_download(client, api_key, torrent_id, selected_file_id)
            if final_url:
                logger.info(
                    "TorBox resolved playback | hash=%s | torrent=%s | file=%s | candidates=%s",
                    hash_val,
                    torrent_id,
                    selected_file_id,
                    len(candidate_ids),
                )
                return final_url

        logger.warning(
            "TorBox all candidate files failed | hash=%s | torrent=%s | candidates=%s | state=%s | progress=%s",
            hash_val,
            torrent_id,
            ",".join(candidate_ids),
            target.get("download_state"),
            target.get("progress"),
        )
    except Exception:
        logger.exception("TorBox resolve error su hash %s", hash_val)
    return None
